src/rust/runtime/memory/buffer/demibuffer.rs

```diff
@@ -0,0 +1,820 @@
+    next: Option<NonNull<MetaData>>,
+
+    // Various fields for TX offload.
+    _tx_offload: u64,
+
+    // Pointer to shared info (rte_mbuf_ext_shared_info).  DPDK uses this for external MBufs.
+    _shinfo: u64,
+
+    // Size of private data (between rte_mbuf struct and the data) in direct MBufs.
+    _priv_size: u16,
+    // Timesync flags for use with IEEE 1588 "Precision Time Protocol" (PTP).
+    _timesync: u16,
+    // Reserved for dynamic fields.
+    _dynfield: [u32; 9],
+}
+
+// MetaData "offload flags".  These exactly mimic those of DPDK MBufs.
+
+// Indicates this MetaData struct doesn't have the actual data directly attached, but rather this MetaData's buf_addr
+// points to another MetaData's directly attached data.
+const METADATA_F_INDIRECT: u64 = 1 << 62;
+
+impl MetaData {
+    // Note on Reference Counts:
+    // Since we are currently single-threaded, there is no need to use atomic operations for refcnt manipulations.
+    // We should rework the implementation of inc_refcnt() and dec_refcnt() to use atomic operations if this changes.
+    // Also, we intentionally don't check for refcnt overflow.  This matches DPDK's behavior, which doesn't check for
+    // reference count overflow either (we're highly unlikely to ever have 2^16 copies of the same data).
+
+    // Increments the reference count and returns the new value.
+    #[inline]
+    pub fn inc_refcnt(&mut self) -> u16 {
+        self.refcnt += 1;
+        self.refcnt
+    }
+
+    // Decrements the reference count and returns the new value.
+    #[inline]
+    pub fn dec_refcnt(&mut self) -> u16 {
+        self.refcnt -= 1;
+        self.refcnt
+    }
+
+    // Gets the MetaData for the last segment in the buffer chain.
+    #[inline]
+    pub fn get_last_segment(&mut self) -> &mut MetaData {
+        let mut md: &mut MetaData = self;
+        while md.next.is_some() {
+            // Safety: The call to as_mut is safe, as the pointer is aligned and dereferenceable, and the MetaData
+            // struct it points to is initialized properly.
+            md = unsafe { md.next.unwrap().as_mut() };
+        }
+        &mut *md
+    }
+}
+
+// The DemiBuffer.
+pub struct DemiBuffer {
+    // Pointer to the buffer metadata.
+    // Stored as a NonNull so it can efficiently be packed into an Option.
+    // This is a "tagged pointer" where the lower bits encode the type of buffer this points to.
+    ptr: NonNull<MetaData>,
+    // Hint to compiler that this struct "owns" a MetaData (for safety determinations).  Doesn't consume space.
+    _phantom: PhantomData<MetaData>,
+}
+
+// DemiBuffer tag types.
+// Since our MetaData structure is 64-byte aligned, the lower 6 bits of a pointer to it are guaranteed to be zero.
+// We currently only use the lower 2 of those bits to hold the tag.
+const TAG_MASK: usize = 0x3;
+const TAG_HEAP: usize = 0x1;
+const TAG_DPDK: usize = 0x2;
```

Comment:
Could you please use an Enumerated Type for this? It will guarantee the right semantics and we can provide the `from` methods.

Reply:
I originally tried that, but it was problematic for bitwise ORs and ANDs, so I switched to constants.  I'll look into this again to see if it can be done cleanly in Rust.  If it was a proper Rust enum, then I wouldn't need the match "catch-all" cases that you complain about below.

Reply:
It is possible to do it, but you need to provide some trait implementations.

Reply:
I switched the valid tag values to a new Tag enum.  The TAG_MASK const stays, as it is only used for bitwise operations.  This allows match operations on Tags to be exhaustive, so no "catch-all" is needed.  Needed to implement BitOr and From traits for my use cases.

Wondering if I should have implemented TryFrom instead of From, and put the panic on memory corruption into get_tag instead.